server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -634,23 +667,23 @@ private void addMetaData(MetaData metaData) throws IOException {
             }
         }
 
-        public void commit(long currentTerm, long lastAcceptedVersion) {
+        public void commit(long currentTerm, long lastAcceptedVersion) throws IOException {
+            ensureOpen();
             try {
                 for (MetaDataIndexWriter metaDataIndexWriter : metaDataIndexWriters) {
                     metaDataIndexWriter.commit(nodeId, currentTerm, lastAcceptedVersion);
                 }
-            } catch (IOException e) {
-                // The commit() call has similar semantics to a fsync(): although it's atomic, if it fails then we've no idea whether the
-                // data on disk is now the old version or the new version, and this is a disaster. It's safest to fail the whole node and
-                // retry from the beginning.
-                throw new IOError(e);
+            } finally {
+                closeIfAnyIndexWriterHasTragedyOrIsClosed();
```

Comment:
Continuing here seems risky -- at least, it's outside what we have modelled for correctness. If we successfully write the data to disk then fail (e.g. unrelated fsync failure) then continue then we will treat that as if we didn't update the last-accepted state and maybe emit some further messages based on that older state, but a reboot will pick up the newer state.

Reply:
We may be able to add some resilience by calling `IndexWriter#prepareCommit()` and handling an exception from that gracefully, since an exception while preparing the commit doesn't deviate from the model.

Reply:
What do you think about alternatively failing calls to the getters (i.e. `getCurrentTerm` and `getLastAcceptedClusterState`) and reloading the writer and forcing the cluster state to disk again if certain amount of time elapsed?

Reply:
IMO this is excessively heroic. A failure in `commit()` indicates something very fundamentally wrong with the node and I think dying is the right thing to do. Note that this is pretty much our behaviour today (since 7.0) and I haven't heard of a single case of this actually happening.

Reply:
ok, I've pushed 7082947b86a